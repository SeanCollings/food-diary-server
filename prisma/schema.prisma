// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// https://www.prisma.io/nestjs
// https://docs.nestjs.com/recipes/prisma
// https://blog.logrocket.com/how-to-use-nestjs-prisma/

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model User {
  id        Int        @id @default(autoincrement())
  name      String
  email     String     @unique
  password  String
  avatar    String?
  createdAt DateTime   @default(now())
  lastLogin DateTime   @default(now())
  shareLink ShareLink?
  diaryDay  DiaryDay[]

  userPreferenceShowDayStreak       Boolean  @default(true)
  userPreferenceShowWeeklyExcercise Boolean?
  userPreferenceShowWeeklyWater     Boolean?

  statLastActivity DateTime?
  statDayStreak    Int?
}

model ShareLink {
  id       Int     @id @default(autoincrement())
  user     User    @relation(fields: [userId], references: [id])
  userId   Int     @unique
  link     String  @unique
  isShared Boolean
}

model DiaryDay {
  id                       String  @id @default(uuid())
  user                     User    @relation(fields: [userId], references: [id])
  userId                   Int
  date                     String
  mealBreakfast            Json?
  mealSnack1               Json?
  mealLunch                Json?
  mealSnack2               Json?
  mealDinner               Json?
  wellnessWater            Float?
  wellnessTeaCoffee        Float?
  wellnessAlcohol          Float?
  wellnessExcercise        String?
  wellnessExcerciseDetails String?

  hasMealBreakfast     Boolean?
  hasMealSnack1        Boolean?
  hasMealLunch         Boolean?
  hasMealSnack2        Boolean?
  hasMealDinner        Boolean?
  hasWellnessExcercise Boolean?

  @@unique([userId, date])
}
