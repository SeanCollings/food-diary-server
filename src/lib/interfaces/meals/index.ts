export enum EMealType {
  BREAKFAST = 'breakfast',
  SNACK_1 = 'snack_1',
  LUNCH = 'lunch',
  SNACK_2 = 'snack_2',
  DINNER = 'dinner',
}
export type TMealType = `${EMealType}`;

export type TSelectedEmoji = {
  name: string;
  nativeSkin: string;
} | null;

export interface IMealContent {
  id: string;
  emoji?: TSelectedEmoji;
  serving?: string;
  measurement?: string;
  food: string;
  description?: string;
}

export interface IMealTypeContent {
  contents: IMealContent[];
}

export type TMealCard = {
  [key in TMealType]?: IMealTypeContent;
};
