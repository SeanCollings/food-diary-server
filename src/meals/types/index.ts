import { IsDateString, IsNotEmpty } from 'class-validator';
import { Request } from 'express';

// https://stackoverflow.com/questions/50160662/convert-nested-array-type

export class CreateMealEntryQuery {
  @IsNotEmpty()
  @IsDateString()
  date: string;
}

export class UpdateMealEntryQuery {
  @IsNotEmpty()
  @IsDateString()
  date: string;
}

export class DeleteMealEntryQuery {
  @IsNotEmpty()
  @IsDateString()
  date: string;
}

export enum MealTypesEnum {
  BREAKFAST = 'breakfast',
  SNACK_1 = 'snack_1',
  LUNCH = 'lunch',
  SNACK_2 = 'snack_2',
  DINNER = 'dinner',
}
export type MealTypes = `${MealTypesEnum}`;

export interface RequestWithUser extends Request {
  user: { userId: string; email: string };
}

export interface MealContent {
  id: string;
  food: string;
  serving?: string;
  quantity?: string;
  description?: string;
  emoji?: { name: string; nativeSkin: string } | null;
}

export type SetFieldOld<V, O extends object> = {
  [K in keyof O]: O[K] extends object ? SetFieldOld<V, O[K]> : V;
};

type Primitive = string | number | boolean | null | undefined;

type SetField<V, O> = O extends Primitive
  ? V
  : O extends (infer U)[]
  ? SetField<V, U>
  : { [K in keyof O]: SetField<V, O[K]> };

type RequireOne<T> = T & { [P in keyof T]: Required<Pick<T, P>> }[keyof T];
