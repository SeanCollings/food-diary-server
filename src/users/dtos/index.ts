export * from '@/users/dtos/update-user.dto';
export * from '@/users/dtos/update-preferences.dto';
export * from '@/users/dtos/user.dto';
