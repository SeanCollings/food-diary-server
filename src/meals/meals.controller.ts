import { JwtAuthGuard } from '@/auth/jwt-auth.guard';
import {
  Body,
  Controller,
  Delete,
  Post,
  Put,
  Query,
  Request,
  UseGuards,
  InternalServerErrorException,
} from '@nestjs/common';
import { MealsService } from '@/meals/meals.service';
import { CreateMealItemDTO } from '@/meals/dtos/create-meal-item.dto';
import {
  CreateMealEntryQuery,
  DeleteMealEntryQuery,
  RequestWithUser,
  UpdateMealEntryQuery,
} from '@/meals/types';
import { UpdateMealItemDto } from '@/meals/dtos/update-meal-item.dto';
import { DeleteMealItemDto } from './dtos/delete-meal-item.dto';
import { DEFAULT_ERROR_MSG } from '@/lib/validation/validation.constants';

@Controller('meals')
export class MealsController {
  constructor(private mealsService: MealsService) {}

  @Post('/')
  @UseGuards(JwtAuthGuard)
  async createMealEntry(
    @Request() req: RequestWithUser,
    @Query() query: CreateMealEntryQuery,
    @Body() body: CreateMealItemDTO,
  ) {
    try {
      return await this.mealsService.createMealEntry(
        req.user.userId,
        query.date,
        body,
      );
    } catch (err) {
      const errorMessage = err.message || DEFAULT_ERROR_MSG;
      console.error('[meals::_post_]:', errorMessage);
      throw new InternalServerErrorException(errorMessage);
    }
  }

  @Put('/')
  @UseGuards(JwtAuthGuard)
  async updateMealEntry(
    @Request() req: RequestWithUser,
    @Query() query: UpdateMealEntryQuery,
    @Body() body: UpdateMealItemDto,
  ) {
    try {
      return await this.mealsService.updateMealEntry(
        req.user.userId,
        query.date,
        body,
      );
    } catch (err) {
      const errorMessage = err.message || DEFAULT_ERROR_MSG;
      console.error('[meals::_put_]:', errorMessage);
      throw new InternalServerErrorException(errorMessage);
    }
  }

  @Delete('/')
  @UseGuards(JwtAuthGuard)
  async deleteMealEntry(
    @Request() req: RequestWithUser,
    @Query() query: DeleteMealEntryQuery,
    @Body() body: DeleteMealItemDto,
  ) {
    try {
      return await this.mealsService.deleteMealEntry(
        req.user.userId,
        query.date,
        body,
      );
    } catch (err) {
      const errorMessage = err.message || DEFAULT_ERROR_MSG;
      console.error('[meals::_delete_]:', errorMessage);
      throw new InternalServerErrorException(errorMessage);
    }
  }
}
