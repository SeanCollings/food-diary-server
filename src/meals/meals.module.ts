import { PrismaService } from '@/prisma.service';
import { Module } from '@nestjs/common';
import { MealsController } from '@/meals/meals.controller';
import { MealsService } from '@/meals/meals.service';

@Module({
  controllers: [MealsController],
  providers: [MealsService, PrismaService],
})
export class MealsModule {}
