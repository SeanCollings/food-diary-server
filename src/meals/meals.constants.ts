import { MealTypesEnum } from '@/meals/types';

export const MEAL_TYPE_ARRAY = Object.values(MealTypesEnum);
