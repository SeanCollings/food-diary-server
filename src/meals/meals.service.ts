import { Injectable } from '@nestjs/common';
import { CreateMealItemDTO } from '@/meals/dtos/create-meal-item.dto';
import { PrismaService } from '@/prisma.service';
import { formatToServerDate } from '@/utils/date-utils';
import { UpdateMealItemDto } from './dtos/update-meal-item.dto';
import { DeleteMealItemDto } from './dtos/delete-meal-item.dto';
import { UsersService } from '@/users/users.service';
import {
  buildMealContent,
  getMealContents,
  mealsColumnLookup,
  removeMealContent,
  updateMealContent,
} from '@/utils/modules/meals-utils';

@Injectable()
export class MealsService {
  constructor(
    private prisma: PrismaService,
    private usersService: UsersService,
  ) {}

  async createMealEntry(userId: number, date: string, meal: CreateMealItemDTO) {
    const serverDate = formatToServerDate(date);
    const [mealColumnName, hasMealContentColumn] = mealsColumnLookup(
      meal.mealId,
    );

    const currentDiaryDay = await this.prisma.diaryDay.findUnique({
      where: {
        userId_date: {
          date: serverDate,
          userId,
        },
      },
    });

    const currentMealContent = getMealContents(currentDiaryDay, mealColumnName);
    const mealContent = buildMealContent(currentMealContent, meal);

    await this.prisma.diaryDay.upsert({
      where: {
        userId_date: {
          date: serverDate,
          userId,
        },
      },
      create: {
        userId,
        date: serverDate,
        [mealColumnName]: mealContent,
        [hasMealContentColumn]: true,
      },
      update: {
        [mealColumnName]: mealContent,
        [hasMealContentColumn]: true,
      },
    });

    return this.usersService.updateUserStreak(userId);
  }

  async updateMealEntry(userId: number, date: string, meal: UpdateMealItemDto) {
    const serverDate = formatToServerDate(date);
    const [newMealColumnName, newHasMealContentColumn] = mealsColumnLookup(
      meal.newMealId,
    );
    const [currentMealColumnName, currentHasMealContentColumn] =
      mealsColumnLookup(meal.oldMealId);

    const currentDiaryDay = await this.prisma.diaryDay.findUnique({
      where: {
        userId_date: {
          date: serverDate,
          userId,
        },
      },
    });

    if (!currentDiaryDay) {
      return;
    }

    const currentMealContents = getMealContents(
      currentDiaryDay,
      currentMealColumnName,
    );
    const newMealContents = getMealContents(currentDiaryDay, newMealColumnName);
    const updatedContent = updateMealContent({
      content: meal.content,
      currentMealColumnName,
      newMealColumnName,
      currentHasMealContentColumn,
      newHasMealContentColumn,
      newMealContents,
      currentMealContents,
    });

    await this.prisma.diaryDay.update({
      where: {
        userId_date: {
          date: serverDate,
          userId,
        },
      },
      data: {
        ...updatedContent,
      },
    });

    return this.usersService.updateUserStreak(userId);
  }

  async deleteMealEntry(userId: number, date: string, meal: DeleteMealItemDto) {
    const serverDate = formatToServerDate(date);
    const [mealColumnName, hasMealContentColumn] = mealsColumnLookup(
      meal.mealId,
    );

    const currentDiaryDay = await this.prisma.diaryDay.findUnique({
      where: {
        userId_date: {
          date: serverDate,
          userId,
        },
      },
    });

    if (!currentDiaryDay) {
      return;
    }

    const currentMealContent = getMealContents(currentDiaryDay, mealColumnName);
    const updatedContent = removeMealContent(currentMealContent, meal);

    await this.prisma.diaryDay.update({
      where: {
        userId_date: {
          date: serverDate,
          userId,
        },
      },
      data: {
        [mealColumnName]: updatedContent,
        [hasMealContentColumn]: !!updatedContent.length,
      },
    });

    return this.usersService.updateUserStreak(userId);
  }
}
