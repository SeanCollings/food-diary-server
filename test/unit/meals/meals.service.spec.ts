import { CreateMealItemDTO } from '@/meals/dtos/create-meal-item.dto';
import { DeleteMealItemDto } from '@/meals/dtos/delete-meal-item.dto';
import { UpdateMealItemDto } from '@/meals/dtos/update-meal-item.dto';
import { MealsService } from '@/meals/meals.service';
import { PrismaService } from '@/prisma.service';
import { UsersService } from '@/users/users.service';
import { Test, TestingModule } from '@nestjs/testing';
import { DiaryDay, PrismaClient } from '@prisma/client';
import { DeepMockProxy, mockDeep } from 'jest-mock-extended';

jest.mock('@/utils/date-utils', () => ({
  ...jest.requireActual('@/utils/date-utils'),
  formatToServerDate: jest.fn().mockReturnValue('2023-04-13'),
}));

describe('MealsService', () => {
  const mockUserId = 'mock_user_id';
  const mockDate = '2023-04-13';
  const mockCreateMealItemDto: CreateMealItemDTO = {
    mealId: 'breakfast',
    content: {
      id: '2',
      food: 'mock_food_2',
      description: 'mock_description_2',
      measurement: 'mock_measurement_2',
      serving: 'mock_serving_2',
      emoji: {
        name: 'mock_emoji_name_2',
        nativeSkin: 'mock_emoji_native_skin_2',
      },
    },
  };
  const mockOtherTypeCreateDto: CreateMealItemDTO = {
    mealId: 'dinner',
    content: {
      id: '3',
      food: 'other_food_type_3',
      description: 'other_food_description_2',
    },
  };

  const mockDiaryDayResult: Partial<DiaryDay> = {
    id: '1',
    date: mockDate,
    userId: mockUserId,
    mealBreakfast: [
      { id: '1', food: 'mock_food_1', description: 'mock_description_2' },
    ],
    hasMealBreakfast: true,
  };

  let service: MealsService;
  let prisma: DeepMockProxy<PrismaClient>;

  const mockUsersService: jest.Mocked<Partial<UsersService>> = {
    updateUserStreak: jest.fn(),
  };

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [MealsService, UsersService, PrismaService],
    })
      .overrideProvider(PrismaService)
      .useValue(mockDeep<PrismaClient>())
      .overrideProvider(UsersService)
      .useValue(mockUsersService)
      .compile();

    service = module.get<MealsService>(MealsService);
    prisma = module.get(PrismaService);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('createMealEntry', () => {
    it('should create a new meal entry and update user streak when there is no current diary entry', async () => {
      prisma.diaryDay.findUnique.mockResolvedValue(null);
      prisma.diaryDay.upsert.mockResolvedValue({} as any);

      await service.createMealEntry(
        mockUserId,
        mockDate,
        mockCreateMealItemDto,
      );

      expect(prisma.diaryDay.findUnique).toHaveBeenCalled();
      expect(prisma.diaryDay.upsert.mock.calls[0][0]).toMatchSnapshot();
      expect(mockUsersService.updateUserStreak).toHaveBeenCalledWith(
        mockUserId,
      );
    });

    it('should update current meal entry type if exists and update user streak', async () => {
      prisma.diaryDay.findUnique.mockResolvedValue(mockDiaryDayResult as any);
      prisma.diaryDay.upsert.mockResolvedValue({} as any);

      await service.createMealEntry(
        mockUserId,
        mockDate,
        mockCreateMealItemDto,
      );

      expect(prisma.diaryDay.findUnique).toHaveBeenCalled();
      expect(prisma.diaryDay.upsert.mock.calls[0][0]).toMatchSnapshot();
      expect(mockUsersService.updateUserStreak).toHaveBeenCalledWith(
        mockUserId,
      );
    });

    it('should create new meal entry if type doesnt exist but diary day entry does', async () => {
      prisma.diaryDay.findUnique.mockResolvedValue(mockDiaryDayResult as any);
      prisma.diaryDay.upsert.mockResolvedValue({} as any);

      await service.createMealEntry(
        mockUserId,
        mockDate,
        mockOtherTypeCreateDto,
      );

      expect(prisma.diaryDay.findUnique).toHaveBeenCalled();
      expect(prisma.diaryDay.upsert.mock.calls[0][0]).toMatchSnapshot();
      expect(mockUsersService.updateUserStreak).toHaveBeenCalledWith(
        mockUserId,
      );
    });
  });

  describe('updateMealEntry', () => {
    const mockUpdateMealItemDto: UpdateMealItemDto = {
      ...mockCreateMealItemDto,
      oldMealId: 'breakfast',
      newMealId: 'dinner',
    };
    const mockUpdateSameMealItemDto: UpdateMealItemDto = {
      ...mockCreateMealItemDto,
      oldMealId: 'breakfast',
      newMealId: 'breakfast',
    };
    const mockUpdateMealNoExist: UpdateMealItemDto = {
      ...mockCreateMealItemDto,
      oldMealId: 'snack_1',
      newMealId: 'snack_2',
    };

    beforeEach(() => {
      prisma.diaryDay.findUnique.mockResolvedValue(mockDiaryDayResult as any);
      prisma.diaryDay.update.mockResolvedValue({} as any);
    });

    it('should update a current meal contents to a new type', async () => {
      await service.updateMealEntry(
        mockUserId,
        mockDate,
        mockUpdateMealItemDto,
      );

      expect(prisma.diaryDay.findUnique).toHaveBeenCalled();
      expect(prisma.diaryDay.update.mock.calls[0][0]).toMatchSnapshot();
      expect(mockUsersService.updateUserStreak).toHaveBeenCalledWith(
        mockUserId,
      );
    });

    it('should update a current meal contents with new meal contents if meal types are the same', async () => {
      await service.updateMealEntry(
        mockUserId,
        mockDate,
        mockUpdateSameMealItemDto,
      );

      expect(prisma.diaryDay.findUnique).toHaveBeenCalled();
      expect(prisma.diaryDay.update.mock.calls[0][0]).toMatchSnapshot();
      expect(mockUsersService.updateUserStreak).toHaveBeenCalledWith(
        mockUserId,
      );
    });

    it('should cater for if current meal doesnt exist', async () => {
      await service.updateMealEntry(
        mockUserId,
        mockDate,
        mockUpdateMealNoExist,
      );

      expect(prisma.diaryDay.findUnique).toHaveBeenCalled();
      expect(prisma.diaryDay.update.mock.calls[0][0]).toMatchSnapshot();
      expect(mockUsersService.updateUserStreak).toHaveBeenCalledWith(
        mockUserId,
      );
    });

    it('should cater for diary-day not existing and not update', async () => {
      prisma.diaryDay.findUnique.mockResolvedValue(null);

      await service.updateMealEntry(
        mockUserId,
        mockDate,
        mockUpdateMealItemDto,
      );

      expect(prisma.diaryDay.update).not.toBeCalled();
      expect(mockUsersService.updateUserStreak).not.toBeCalled();
    });
  });

  describe('deleteMealEntry', () => {
    const deleteDto: DeleteMealItemDto = {
      id: '1',
      mealId: 'breakfast',
    };
    const deleteNoIdDto: DeleteMealItemDto = {
      id: '1000000',
      mealId: 'breakfast',
    };
    const deleteNoTypeDto: DeleteMealItemDto = {
      id: '1',
      mealId: 'snack_1',
    };

    beforeEach(() => {
      prisma.diaryDay.findUnique.mockResolvedValue(mockDiaryDayResult as any);
      prisma.diaryDay.update.mockResolvedValue({} as any);
    });

    it('should remove a meal by id from current meals', async () => {
      await service.deleteMealEntry(mockUserId, mockDate, deleteDto);

      expect(prisma.diaryDay.findUnique).toHaveBeenCalled();
      expect(prisma.diaryDay.update.mock.calls[0][0]).toMatchSnapshot();
      expect(mockUsersService.updateUserStreak).toHaveBeenCalledWith(
        mockUserId,
      );
    });

    it('should cater for if trying to remove id that doesnt exist', async () => {
      await service.deleteMealEntry(mockUserId, mockDate, deleteNoIdDto);

      expect(prisma.diaryDay.findUnique).toHaveBeenCalled();
      expect(prisma.diaryDay.update.mock.calls[0][0]).toMatchSnapshot();
      expect(mockUsersService.updateUserStreak).toHaveBeenCalledWith(
        mockUserId,
      );
    });

    it('should cater for if trying to remove meal-type that doesnt exist', async () => {
      await service.deleteMealEntry(mockUserId, mockDate, deleteNoTypeDto);

      expect(prisma.diaryDay.findUnique).toHaveBeenCalled();
      expect(prisma.diaryDay.update.mock.calls[0][0]).toMatchSnapshot();
      expect(mockUsersService.updateUserStreak).toHaveBeenCalledWith(
        mockUserId,
      );
    });

    it('should cater for diary-day not existing and not update', async () => {
      prisma.diaryDay.findUnique.mockResolvedValue(null);

      await service.deleteMealEntry(mockUserId, mockDate, deleteDto);

      expect(prisma.diaryDay.update).not.toBeCalled();
      expect(mockUsersService.updateUserStreak).not.toBeCalled();
    });
  });
});
