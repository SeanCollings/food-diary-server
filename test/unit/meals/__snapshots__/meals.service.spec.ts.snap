// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MealsService createMealEntry should create a new meal entry and update user streak when there is no current diary entry 1`] = `
Object {
  "create": Object {
    "date": "2023-04-12T22:00:00.000Z",
    "hasMealBreakfast": true,
    "mealBreakfast": Array [
      Object {
        "description": "mock_description_2",
        "emoji": Object {
          "name": "mock_emoji_name_2",
          "nativeSkin": "mock_emoji_native_skin_2",
        },
        "food": "mock_food_2",
        "id": "2",
        "measurement": "mock_measurement_2",
        "serving": "mock_serving_2",
      },
    ],
    "userId": 1234,
  },
  "update": Object {
    "hasMealBreakfast": true,
    "mealBreakfast": Array [
      Object {
        "description": "mock_description_2",
        "emoji": Object {
          "name": "mock_emoji_name_2",
          "nativeSkin": "mock_emoji_native_skin_2",
        },
        "food": "mock_food_2",
        "id": "2",
        "measurement": "mock_measurement_2",
        "serving": "mock_serving_2",
      },
    ],
  },
  "where": Object {
    "userId_date": Object {
      "date": "2023-04-12T22:00:00.000Z",
      "userId": 1234,
    },
  },
}
`;

exports[`MealsService createMealEntry should create new meal entry if type doesnt exist but diary day entry does 1`] = `
Object {
  "create": Object {
    "date": "2023-04-12T22:00:00.000Z",
    "hasMealDinner": true,
    "mealDinner": Array [
      Object {
        "description": "other_food_description_2",
        "food": "other_food_type_3",
        "id": "3",
      },
    ],
    "userId": 1234,
  },
  "update": Object {
    "hasMealDinner": true,
    "mealDinner": Array [
      Object {
        "description": "other_food_description_2",
        "food": "other_food_type_3",
        "id": "3",
      },
    ],
  },
  "where": Object {
    "userId_date": Object {
      "date": "2023-04-12T22:00:00.000Z",
      "userId": 1234,
    },
  },
}
`;

exports[`MealsService createMealEntry should update current meal entry type if exists and update user streak 1`] = `
Object {
  "create": Object {
    "date": "2023-04-12T22:00:00.000Z",
    "hasMealBreakfast": true,
    "mealBreakfast": Array [
      Object {
        "description": "mock_description_2",
        "food": "mock_food_1",
        "id": "1",
      },
      Object {
        "description": "mock_description_2",
        "emoji": Object {
          "name": "mock_emoji_name_2",
          "nativeSkin": "mock_emoji_native_skin_2",
        },
        "food": "mock_food_2",
        "id": "2",
        "measurement": "mock_measurement_2",
        "serving": "mock_serving_2",
      },
    ],
    "userId": 1234,
  },
  "update": Object {
    "hasMealBreakfast": true,
    "mealBreakfast": Array [
      Object {
        "description": "mock_description_2",
        "food": "mock_food_1",
        "id": "1",
      },
      Object {
        "description": "mock_description_2",
        "emoji": Object {
          "name": "mock_emoji_name_2",
          "nativeSkin": "mock_emoji_native_skin_2",
        },
        "food": "mock_food_2",
        "id": "2",
        "measurement": "mock_measurement_2",
        "serving": "mock_serving_2",
      },
    ],
  },
  "where": Object {
    "userId_date": Object {
      "date": "2023-04-12T22:00:00.000Z",
      "userId": 1234,
    },
  },
}
`;

exports[`MealsService deleteMealEntry should cater for if trying to remove id that doesnt exist 1`] = `
Object {
  "data": Object {
    "hasMealBreakfast": true,
    "mealBreakfast": Array [
      Object {
        "description": "mock_description_2",
        "food": "mock_food_1",
        "id": "1",
      },
    ],
  },
  "where": Object {
    "userId_date": Object {
      "date": "2023-04-12T22:00:00.000Z",
      "userId": 1234,
    },
  },
}
`;

exports[`MealsService deleteMealEntry should cater for if trying to remove meal-type that doesnt exist 1`] = `
Object {
  "data": Object {
    "hasMealSnack1": false,
    "mealSnack1": Array [],
  },
  "where": Object {
    "userId_date": Object {
      "date": "2023-04-12T22:00:00.000Z",
      "userId": 1234,
    },
  },
}
`;

exports[`MealsService deleteMealEntry should remove a meal by id from current meals 1`] = `
Object {
  "data": Object {
    "hasMealBreakfast": false,
    "mealBreakfast": Array [],
  },
  "where": Object {
    "userId_date": Object {
      "date": "2023-04-12T22:00:00.000Z",
      "userId": 1234,
    },
  },
}
`;

exports[`MealsService updateMealEntry should cater for if current meal doesnt exist 1`] = `
Object {
  "data": Object {
    "hasMealSnack1": false,
    "hasMealSnack2": true,
    "mealSnack1": Array [],
    "mealSnack2": Array [
      Object {
        "description": "mock_description_2",
        "emoji": Object {
          "name": "mock_emoji_name_2",
          "nativeSkin": "mock_emoji_native_skin_2",
        },
        "food": "mock_food_2",
        "id": "2",
        "measurement": "mock_measurement_2",
        "serving": "mock_serving_2",
      },
    ],
  },
  "where": Object {
    "userId_date": Object {
      "date": "2023-04-12T22:00:00.000Z",
      "userId": 1234,
    },
  },
}
`;

exports[`MealsService updateMealEntry should update a current meal contents to a new type 1`] = `
Object {
  "data": Object {
    "hasMealBreakfast": true,
    "hasMealDinner": true,
    "mealBreakfast": Array [
      Object {
        "description": "mock_description_2",
        "food": "mock_food_1",
        "id": "1",
      },
    ],
    "mealDinner": Array [
      Object {
        "description": "mock_description_2",
        "emoji": Object {
          "name": "mock_emoji_name_2",
          "nativeSkin": "mock_emoji_native_skin_2",
        },
        "food": "mock_food_2",
        "id": "2",
        "measurement": "mock_measurement_2",
        "serving": "mock_serving_2",
      },
    ],
  },
  "where": Object {
    "userId_date": Object {
      "date": "2023-04-12T22:00:00.000Z",
      "userId": 1234,
    },
  },
}
`;

exports[`MealsService updateMealEntry should update a current meal contents with new meal contents if meal types are the same 1`] = `
Object {
  "data": Object {
    "mealBreakfast": Array [
      Object {
        "description": "mock_description_2",
        "food": "mock_food_1",
        "id": "1",
      },
    ],
  },
  "where": Object {
    "userId_date": Object {
      "date": "2023-04-12T22:00:00.000Z",
      "userId": 1234,
    },
  },
}
`;
